pkg/kv/kvserver/kvstorage/init.go

```diff
@@ -356,37 +398,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
 		return nil, err
 	}
 
+	// INVARIANT: all replicas have a persisted full replicaID (i.e. a "replicaID from disk").
+	//
+	// This invariant is true for replicas created in 22.2. Without further action, it
+	// maybe be violated for clusters that originated before 22.2.
+	//
+	// loadFullReplicaIDsFromDisk returns a zero ReplicaID when none is found.
+	//
+	// In this method, we migrate to always satisfy the invariant:
+	//
+	// 1. for initialized replicas, use ReplicaID from RangeDescriptor as RaftReplicaID
+	// 2. remove uninitialized replicas lacking a RaftReplicaID.
```

Comment:
(2) doesn't look like invariant. Did you mean to say that "uninitialized replicas don't have a stored RaftReplicaID"? This isn't yet true either, because `getOrCreateReplica` still writes one asap.

Should we hold off deleting uninit replica IDs from storage until (or after) a PR that would stop writing them too? Otherwise the code will be in a state where we always write them, and remove on restarts, which is a bit wasteful.

Reply:
I misunderstood the comment. (1) and (2) describe migration actions, which are needed to arrive at the invariant above. I might be wrong in my interpretation of what this does, will take a fresh look tomorrow.

Reply:
Ok, so we're looking at 3 KVs of a `Replica`: `RaftReplicaID`, `HardState`, and `RangeDescriptor`.

<table>
<tr>
	<td> RaftReplicaID
	<td> HardState
	<td> RangeDescriptor
	<td> What?
<tr>
	<td> no
	<td> no
	<td> no
	<td> nothing
<tr>
	<td> no
	<td> no
	<td> yes
	<td> -
<tr>
	<td> no
	<td> yes
	<td> no
	<td> uninitialized (v <= 22.1)
<tr>
	<td> no
	<td> yes
	<td> yes
	<td> initialized (v <= 22.1)
<tr>
	<td> yes
	<td> no
	<td> no
	<td> -
<tr>
	<td> yes
	<td> no
	<td> yes
	<td> -
<tr>
	<td> yes
	<td> yes
	<td> no
	<td> uninitialized (v >= 22.2)
<tr>
	<td> yes
	<td> yes
	<td> yes
	<td> initialized
</table>

Steps 1 and 2 in this comment say that we eliminate the "RaftReplicaID: no" half of this table. The way step 1 and 2 do so is different though, because for uninit replica we don't know the Replica ID; if we did we would write it too; but instead we remove such replicas.

Makes sense.

Reply:
It's a bit hard to follow all the map maths below. I will try now, but I think this could be simpler if we just implemented something like a table that a drew above:

```golang
type rangeState struct {
	replicaID roachpb.ReplicaID
	hardState raftpb.HardState
	rangeDesc roachpb.RangeDescriptor
}

var m map[roachpb.RangeID]rangeState

read all RaftReplicaID
insert into m[rangeID].replicaID (init a new rangeState if it does not exist)

read all HardState
insert into m[rangeID].hardState (init a new rangeState if it does not exist)

read all RangeDescriptor
insert into m[rangeID].rangeDesc (init a new rangeState if it does not exist)

for rangeID, state := range m {
	checkInvariants(rangeID, state) // e.g. assert that we don't see one of the impossible "-" combinations
	writeMigration(rangeID, state)
}
```

Reply:
Good idea, this simplified it a lot. I'll ping you when the review is ready.